src/messaging/Channel.h

```diff
@@ -0,0 +1,195 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the API classes for to CHIP Channel.
+ *
+ *      Channel is a object to abstract all low layer dependencies of an
+ *      exchange, including secure session, transport connection and network
+ *      status.
+ *
+ *      Channel is not connection. Channel can abstract both connection
+ *      oriented and connectionless transports, It contains information to send
+ *      and receive messages via exchanges. For example, when using
+ *      connectionless transport, channel will contain peer address and session
+ *      key; when using connection oriented transport, channel will contain
+ *      connection handle and session key.
+ *
+ *      Channel is not session. Session do persistent through cold reboot, but
+ *      channel doesn't. Applications must re-establish channels after a cold
+ *      reboot.
+ *
+ *      Because channel is a local concept, peer device is not able aware of
+ *      channel establishment events. Instead, peer device is able to aware
+ *      session establishment events, connection establishment events for
+ *      connection oriented transport and message received events for
+ *      connectionless transport.
+ */
+
+#pragma once
+
+#include <transport/raw/MessageHeader.h>
+
+namespace chip {
+
+/**
+ *  @brief
+ *    The ChannelBuilder object provides information to build a Channel.
+ *
+ *    ChannelBuilder can be used by application to provide information to
+ *    request a channel to a peer. When a channel is requested via
+ *    ExchangeManager::EstablishChannel, a ChannelHandle will be return to
+ *    represent the state of the channel
+ *
+ *    The ChannelBuilder object should be short-live and it is only used within
+ *    ExchangeManager::EstablishChannel call, after the call its state will be
+ *    copied into an internal object represented by ChannelHandle, then the
+ *    ChannelBuilder can be safely released.
+ */
+class ChannelBuilder
+{
+public:
+    enum class NetworkPreference
+    {
+        kNetwork_Default,
+        kNetwork_HighBandwidth,
+        kNetwork_LowLatency,
+        kNetwork_LowPowerConsumption,
+    };
+
+    enum class TransportPreference
+    {
+        kTransport_Connectionless,           // default
```

Comment:
should we add a kTransportDefault as well here as an alias to connectionless, o that we do not need the 'default' comment?